### PR TITLE
AS-321: entity types data repo [risk: low]

### PR DESCRIPTION
**Pull request by davidangb**
This PR:
* accepts `dataReference` and `billingProject` query params for the entity metadata api
  * generalized at the code level to accept these params for all entity apis; only used by metadata api
* adds a `DataRepoEntityProvider` and `DataRepoEntityProviderBuilder`
* adds a `DataRepoDAO` trait and `HttpDataRepoDAO` impl
  * implemented via data repo client library
* for entity metadata api, routes to `DataRepoEntityProvider` if `dataReference` query param is nonempty
* `DataRepoEntityProvider` implements:
  * request data reference description from Workspace Manager, based on the `dataReference` name
  * request snapshot description from Terra Data Repo, based on the snapshot id in Workspace Manager's data reference description
  * read tables and columns from Terra Data Repo's snapshot description

### CODE REVIEW

File: project/build.properties
Comment (davidangb) [inline project/build.properties:1]: This PR adds the `datarepo-client` dependency. That has a transitive dependency on `javax.ws.rs:javax.ws.rs-api:2.1.1`. There is a bug in sbt 1.2 - actually in Ivy - that causes that transitive dependency to fail; see https://github.com/sbt/sbt/issues/3618.

Upgrading to sbt 1.3 addresses this problem, because sbt 1.3 uses Coursier, not Ivy. Alternate solution is to ask the data repo team to change their dependency, but the sbt 1.3 upgrade feels worthwhile anyway.

Note that broadinstitute/firecloud-develop#2138 enables coursier caching for local-environment servers. It's not technically required but it avoids re-downloading dependencies every time you start your local server.

EDIT: I've excluded `javax.ws.rs:javax.ws.rs-api:2.1.1` to avoid sbt assembly issues, so this sbt upgrade is no longer necessary ... but I've tested it and it seems worthwhile so I'm leaving it in.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpec.scala
Comment (ungwudik) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpec.scala]: How do we know when TDR starts supporting these? We should consider creating tickets for these tests (if they don't already exist) and linking them to the corresponding ticket on TDR team's JIRA.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala
Comment (ungwudik) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala]: Can these two variables (`wsid` and `refid`) be updated to camelCase for consistency?

File: core/src/main/resources/swagger/data-repo-only.yaml
Comment (MatthewBemis) [inline core/src/main/resources/swagger/data-repo-only.yaml]: shouldn't this also be `in: query`?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala]: i'm most certainly missing the appropriate context, but does Data Repo somehow allow for null PKs or is that comment more related to an unexpected null making it's way into here?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala]: is there a reason to hold off on adding that error-trapping now?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala
Comment (MatthewBemis) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala]: out of an abundance of caution, should this be a little less strict, i.e. `equalsIgnoreCase`?

**Review by MatthewBemis — COMMENTED** [review]
Just a few minor comments!

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala
Comment (davidangb) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala]: I've updated the comment to reflect reality

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpec.scala
Comment (davidangb) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpec.scala]: updated comments.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala
Comment (davidangb) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpecSupport.scala]: done

File: core/src/main/resources/swagger/data-repo-only.yaml
Comment (davidangb) [reply core/src/main/resources/swagger/data-repo-only.yaml]: done

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala
Comment (davidangb) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala]: updated comment

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala
Comment (davidangb) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala]: done

**Comment by davidangb** [comment]
@ungwudik @MatthewBemis ready for re-review!

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpec.scala
Comment (ungwudik) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpec.scala]: Looks like DR-1003 was completed yesterday? I'm not 100% sure what 'Done' means for their team, so adding "tests for entity/row counts returned by data repo" could be a part of a future PR.

**Review by ungwudik — APPROVED** [review]
Looks good to me, pending approvals from Matt and/or Anu.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala
Comment (ansingh7115) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala:62]: I believe by the time we get here, we've asked Sam whether the user has Read access to the workspace. Since snapshots have their own sam permissions - do we need to check that the user also has permissions on the snapshot? Or are we assuming that when a snapshot is associated with a workspace, that means anyone who can or will in the future have Read access to the workspace, should be able to see the snapshot?

Comment (davidangb) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala:62]: @ansingh7115 per PO business requirements, any user who has Read access to the workspace should be able to see all snapshot references in that workspace. This doesn't necessarily mean that the user can see the snapshot *contents* - and those permission checks will happen as soon as we connect to Data Repo. But as far as this specific `lookupSnapshotForName` method - any workspace user should be able to resolve the name into the snapshot UUID.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpec.scala
Comment (davidangb) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpec.scala]: I saw that the jira ticket was marked Done, I did not see actual row count data in the Data Repo dev instance, so I haven't adopted it yet, I'll make a follow-on AS ticket to track this work.

**Review by MatthewBemis — APPROVED** [review]
:+1: pending passing tests

**Comment by davidangb** [comment]
```
write /var/lib/docker/tmp/GetImageBlob612182488: no space left on device
```
in last fiab-start. jenkins retest.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpec.scala
Comment (davidangb) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderSpec.scala]: tracking as AS-365

**Review by ansingh7115 — APPROVED** [review]

**Pull request merged**
